experimental/trusted_shuffler/server/src/http.rs

```diff
@@ -26,26 +26,82 @@ use std::{
     time::Instant,
 };
 use tokio::time::Duration;
-use trusted_shuffler::{RequestHandler, TrustedShuffler};
-use trusted_shuffler_common::send_http_request;
+use trusted_shuffler::{
+    RequestHandler, TrustedShuffler, TrustedShufflerRequest, TrustedShufflerResponse,
+};
+use trusted_shuffler_common::send_with_request;
 
 struct HttpRequestHandler {
     backend_url: String,
 }
 
 #[async_trait]
 impl RequestHandler for HttpRequestHandler {
-    async fn handle(&self, request: Vec<u8>) -> anyhow::Result<Vec<u8>> {
-        let response = send_http_request(&self.backend_url, Method::POST, &request).await;
-        response.map_or_else(
-            |error| {
-                Err(anyhow!(
-                    "Couldn't receive response from the backend: {:?}",
-                    error
-                ))
-            },
-            Ok,
-        )
+    async fn handle(
+        &self,
+        request: TrustedShufflerRequest,
+    ) -> anyhow::Result<TrustedShufflerResponse> {
+        let mut request = trusted_to_hyper_request(request);
+
+        // We want to keep the path of the orginal request from the client.
+        let path = request.uri().path();
+        // And extend the URL to the backend.
+        let new_uri = format!("{}{}", self.backend_url, path)
+            .parse::<Uri>()
+            .expect("Couldn't parse URI for backend.");
+        let uri = request.uri_mut();
+        *uri = new_uri;
+        log::info!("New request to the backend: {:?}\n", request);
+
+        match send_with_request(request).await {
+            Err(error) => Err(anyhow!(
+                "Couldn't receive response from the backend: {:?}",
+                error
+            )),
+            Ok(response) => Ok(hyper_to_trusted_response(response).await),
+        }
+    }
+}
+
+// We keep the body and the uri from the `hyper::Request`. We convert the body to vec, so it has to
+// be read fully.
+async fn hyper_to_trusted_request(hyper_request: Request<Body>) -> TrustedShufflerRequest {
+    let (parts, body) = hyper_request.into_parts();
+    let body = hyper::body::to_bytes(body)
+        .await
+        .expect("Couldn't read request body");
+    TrustedShufflerRequest {
+        body: body.to_vec(),
+        uri: parts.uri.clone(),
+    }
+}
+
+// Apart from setting the body and the URI, for our test backend setting to HTTP/2 and POST
+// seems sufficient.
+fn trusted_to_hyper_request(trusted_shuffler_request: TrustedShufflerRequest) -> Request<Body> {
+    let body = Body::from(trusted_shuffler_request.body);
+    Request::builder()
+        .uri(trusted_shuffler_request.uri)
+        .method(Method::POST)
```

Comment:
(optional) Supporting also arbitrary methods may be an easy win to allow using this same exact binary as a proxy for arbitrary web servers, e.g. it can be pointed in front of wikipedia or some other server and it would make a cool demo from a browser (assuming the web server is HTTP2 compatible). But we can look into that separately.